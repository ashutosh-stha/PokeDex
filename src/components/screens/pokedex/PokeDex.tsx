import React, {useCallback, useEffect, useMemo, useState} from 'react';
import {Button, FlatList, StyleSheet, TextInput, View} from 'react-native';
import {useDispatch, useSelector} from 'react-redux';
import {Dispatch, RootState} from '../../../model/store';
import {Pokemon} from '../../../model/pokemon/pokemon';
import {PokemonItem} from './PokemonItem';
import _ from 'lodash';
import {useNavigation} from '@react-navigation/native';
import {GenericNavigationProps} from '../../../routes/types';
import {POKEMON_DETAIL} from '../../../routes/routes/ScreenRoutes';

export const PokeDex = () => {
  const dispatch = useDispatch<Dispatch>();
  const data = useSelector((state: RootState) => state.pokemon?.pokemonList);

  const [pokemonList, setPokemonList] = useState<Pokemon[]>([]);
  const [selectedPokemon, setSelectedPokemon] = useState<Pokemon[]>([]);
  const [searchTerm, setSearchTerm] = useState('');
  const navigation = useNavigation<GenericNavigationProps>();

  useEffect(() => {
    dispatch.pokemon.loadPokemon({});
  }, [dispatch.pokemon]);

  useEffect(() => {
    setPokemonList(data);
  }, [data]);

  useEffect(() => {
    const debounceSearch = _.debounce(() => {
      setPokemonList(
        data.filter(item => item.name.includes(searchTerm.toLowerCase())),
      );
    }, 300);
    debounceSearch();
    return () => {
      debounceSearch.cancel();
    };
  }, [data, searchTerm]);

  const memoizeButtonPress = useCallback(
    (pokemon: Pokemon) => {
      const selectedList = _.xor(selectedPokemon, [pokemon]);
      setSelectedPokemon(selectedList);
      navigation.navigate(POKEMON_DETAIL, {pokemon: pokemon.name});
    },
    [navigation, selectedPokemon],
  );

  const renderPokemon = (pokemon: Pokemon) => {
    return (
      <PokemonItem
        key={pokemon.name}
        pokemon={pokemon}
        onPress={() => {
          memoizeButtonPress(pokemon);
        }}
        selected={selectedPokemon.includes(pokemon)}
      />
    );
  };

  const onClear = () => {
    setSearchTerm('');
  };

  const memoizedPokemonList = useMemo(() => pokemonList, [pokemonList]);

  return (
    <View style={styles.container}>
      <View style={styles.inputContainer}>
        <TextInput
          testID="searchBox"
          style={styles.textInput}
          value={searchTerm}
          placeholder="Search Pokemon"
          onChangeText={text => setSearchTerm(text)}
        />
        <Button testID="clearBtn" title="Clear" onPress={onClear} />
      </View>
      <FlatList
        testID="pokemonList"
        keyExtractor={item => item.name}
        style={styles.container}
        data={memoizedPokemonList}
        numColumns={2}
        renderItem={({item}) => renderPokemon(item)}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  textInput: {
    borderWidth: 2,
    padding: 10,
    margin: 20,
    borderRadius: 10,
    flex: 0.9,
  },
  inputContainer: {
    alignItems: 'center',
    flexDirection: 'row',
  },
});
