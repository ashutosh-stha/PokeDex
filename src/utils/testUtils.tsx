import React from 'react';
import {Provider} from 'react-redux';
import {Store} from 'redux';
import {init} from '@rematch/core';
import {render} from '@testing-library/react-native';
import {NavigationContainer} from '@react-navigation/native';

type WrapWithStore = {
  children: any;
};
type InitPropsType = {
  wrapper?: any;
};
const wrapWithStore = (store: Store) => {
  return ({children}: WrapWithStore) => (
    <Provider store={store}>
      <NavigationContainer>{children}</NavigationContainer>
    </Provider>
  );
};
class ComponentRenderer {
  private component;
  private models: any;
  private store: any;
  constructor(component: any) {
    this.component = component;
  }

  withModels(models: any) {
    this.models = models;
    return this;
  }
  render() {
    const initProps: InitPropsType = {};
    if (this.models) {
      initProps.wrapper = wrapWithStore(init({models: this.models}));
    }
    return render(this.component, initProps);
  }
}
export function componentRenderer(component: any) {
  return new ComponentRenderer(component);
}
