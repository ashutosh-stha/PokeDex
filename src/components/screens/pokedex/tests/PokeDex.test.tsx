import {PokeDex} from '../PokeDex';
import {componentRenderer} from '../../../../utils/testUtils';
import {fireEvent, screen} from '@testing-library/react-native';
import '@testing-library/jest-native/extend-expect';
import {act} from 'react-test-renderer';

jest.useFakeTimers();

const renderWrapper = () =>
  componentRenderer(<PokeDex />)
    .withModels({
      pokemon: {
        state: {
          pokemonList: [{name: 'Pikachu', url: 'url'}],
        },
        effects: () => ({loadPokemon: jest.fn()}),
      },
    })
    .render();

let wrapper: any;

describe('PokeDex', () => {
  beforeEach(() => {
    wrapper = renderWrapper();
  });
  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('when user types on the search box', () => {
    act(() => {
      jest.advanceTimersByTime(900);
      fireEvent(wrapper.getByTestId('searchBox'), 'onChangeText', 'Ditto');
    });
    expect(wrapper).toMatchSnapshot();
  });
  it('when user press the clear button', () => {
    act(() => {
      fireEvent(wrapper.getByTestId('searchBox'), 'onChangeText', 'Ditto');
      fireEvent.press(screen.getByTestId('clearBtn'));
    });
    expect(screen.getByTestId('searchBox')).toHaveTextContent('');
  });
  it('when user selects the pokemon button', () => {
    act(() => {
      fireEvent.press(screen.getByTestId('pokemonBtn'));
    });
    expect(screen.getByTestId('pokemonBtn')).toHaveStyle({
      backgroundColor: '#262730',
      borderRadius: 30,
      flex: 0.5,
      margin: 8,
      opacity: 1,
      padding: 10,
    });
  });
});
