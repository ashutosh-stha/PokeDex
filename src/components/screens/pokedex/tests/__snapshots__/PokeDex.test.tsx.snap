// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokeDex should render correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Search Pokemon"
      style={
        {
          "borderRadius": 10,
          "borderWidth": 2,
          "flex": 0.9,
          "margin": 20,
          "padding": 10,
        }
      }
      testID="searchBox"
      value=""
    />
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="clearBtn"
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Clear
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    data={
      [
        {
          "name": "Pikachu",
          "url": "url",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    style={
      {
        "flex": 1,
      }
    }
    testID="pokemonList"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#262730",
                "borderRadius": 30,
                "flex": 0.5,
                "margin": 8,
                "opacity": 1,
                "padding": 10,
              }
            }
            testID="pokemonBtn"
          >
            <Text
              style={
                {
                  "color": "white",
                  "fontSize": 16,
                  "textAlign": "center",
                  "textTransform": "capitalize",
                }
              }
            >
              Pikachu
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`PokeDex when user types on the search box 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Search Pokemon"
      style={
        {
          "borderRadius": 10,
          "borderWidth": 2,
          "flex": 0.9,
          "margin": 20,
          "padding": 10,
        }
      }
      testID="searchBox"
      value="Ditto"
    />
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="clearBtn"
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Clear
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    data={
      [
        {
          "name": "Pikachu",
          "url": "url",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    style={
      {
        "flex": 1,
      }
    }
    testID="pokemonList"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#77BA99",
                "borderRadius": 30,
                "flex": 0.5,
                "margin": 8,
                "opacity": 1,
                "padding": 10,
              }
            }
            testID="pokemonBtn"
          >
            <Text
              style={
                {
                  "color": "white",
                  "fontSize": 16,
                  "textAlign": "center",
                  "textTransform": "capitalize",
                }
              }
            >
              Pikachu
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
